### ISSUE TITLE
Cjdns can't create keys in /etc

### ISSUE THREAD

**Issue by unlmtd** [NixOS/nixpkgs#21038]
## Issue description

With cjdns enabled, the service complains that it can't create the necessary
keys in /etc

```/nix/store/9ms47qxfd0qrxn4s689m9pbdrds57d59-unit-script/bin/cjdns-pre-start: line 9: /etc/cjdns.keys: Read-only file system```

### Steps to reproduce

```services.cjdns.enable = true;```

## Technical details

* System: 17.03pre96925.1c50bdd (Gorilla)
* Nix version: 1.11.4
* Nixpkgs version: 17.03pre96925.1c50bdd

**Comment by unlmtd** [comment]
This seems to be the relevant part in nixpkgs/nixos/modules/services/networking/cjdns.nix

```nix
      preStart = if cfg.confFile != null then "" else ''
        [ -e /etc/cjdns.keys ] && source /etc/cjdns.keys
        if [ -z "$CJDNS_PRIVATE_KEY" ]; then
            shopt -s lastpipe
            ${pkg}/bin/makekeys | { read private ipv6 public; }
            umask 0077
            echo "CJDNS_PRIVATE_KEY=$private" >> /etc/cjdns.keys
            echo -e "CJDNS_IPV6=$ipv6\nCJDNS_PUBLIC_KEY=$public" > /etc/cjdns.public
            chmod 600 /etc/cjdns.keys
            chmod 444 /etc/cjdns.public
        fi
        if [ -z "$CJDNS_ADMIN_PASSWORD" ]; then
            echo "CJDNS_ADMIN_PASSWORD=$(tr -dc A-Za-z0-9 </dev/urandom | head -c 96)" \
                >> /etc/cjdns.keys
        fi
      '';
```

What would be the Nixos way here?

**Comment by joachifm** [comment]
Yes, it's because the "hardening" is too zealous, it needs to let preStart run as root (or not run with ProtectSystem = full or whatever)

**Comment by joachifm** [comment]
@unlmtd can you try something like `services.cjdns.serviceConfig.PermissionsStartOnly = true`?

**Comment by unlmtd** [comment]
nixos-rebuild says that option doesnt exist

**Comment by joachifm** [comment]
Agh, sorry, that should be `systemd.services.cjdns.serviceConfig`

**Comment by unlmtd** [comment]
That didn't do it, Im getting the same error.

**Comment by joachifm** [comment]
Hm, I guess it doesn't pertain to fs isolation options; then you need to change `ProtectSystem = "full"` to `true` (or `false`).

**Comment by tohl** [comment]
ProtectSystem = "full" prevents writing the keys (/etc is read-only).

pre-start should be refactored to a separate systemd service without
that restriction, and cjdns run with that restriction in another
service.

**Comment by joachifm** [comment]
This should be fixed by a0338afe5faa9f9e403e2caa52e4a8b60c272be9. A better fix is at https://github.com/joachifm/nixpkgs/tree/cjdns-init-keys but is not very well tested.
